File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	var n int
	var f, res float64
	fmt.Fscanf(reader, "%d\n", &n)
	var nums = make([]float64, n)
	var reverse = make([]float64, n)
	var increase = make([]float64, n)
	var decrease = make([]float64, n)
	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%f ", &f)
		nums[i] = f
		reverse[n-i-1] = f
		increase[i] = 1
		decrease[i] = 1
	}
	for i := 0; i < n; i++ {
		for j := 0; j < i; j++ {
			if nums[i] > nums[j] {
				increase[i] = math.Max(increase[i], increase[j]+1)
			}
			if reverse[i] > reverse[j] {
				decrease[i] = math.Max(decrease[i], decrease[j]+1)
			}
		}
	}
	for i := 0; i < n; i++ {
		res = math.Max(res, increase[i]+decrease[n-i-1]-1)
	}
	fmt.Fprintf(writer, "%0.f\n", res)
}

File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

n = int(input())
nums = list(map(int, input().split()))
reverse = nums[:]
reverse.reverse()
increase = [1 for _ in range(n)]
decrease = [1 for _ in range(n)]
for i in range (n):
    for j in range (i):
        if nums[i]>nums[j]:
            increase[i] = max(increase[i],increase[j]+1)
        if reverse[i]>reverse[j]:
            decrease[i] =max(decrease[i], decrease[j]+1)
res = [0 for _ in range(n)]
for i in range(n):
    res[i] = increase[i] + decrease[n-i-1] -1
print(max(res))